compiler/p/codegen/OMRLoadStoreHandler.cpp

```diff
@@ -0,0 +1,393 @@
+            );
+
+            cg->machine()->setLinkRegisterKilled(true);
+            }
+        else if (cg->comp()->target().cpu.isAtLeast(OMR_PROCESSOR_PPC_P8) && cg->comp()->target().cpu.supportsFeature(OMR_FEATURE_PPC_HAS_VSX))
+            {
+            TR::Register *tempHiReg = cg->allocateRegister(TR_FPR);
+            TR::Register *tempLoReg = cg->allocateRegister(TR_FPR);
+
+            generateTrg1MemInstruction(cg, TR::InstOpCode::lfd, node, tempHiReg, memRef);
+            generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::xxspltw, node, tempLoReg, tempHiReg, 1);
+            generateTrg1Src1Instruction(cg, TR::InstOpCode::mfvsrwz, node, trgReg->getHighOrder(), tempHiReg);
+            generateTrg1Src1Instruction(cg, TR::InstOpCode::mfvsrwz, node, trgReg->getLowOrder(), tempLoReg);
+
+            if (cg->comp()->target().isSMP())
+                generateInstruction(cg, TR::InstOpCode::lwsync, node);
+
+            cg->stopUsingRegister(tempLoReg);
+            cg->stopUsingRegister(tempHiReg);
+            }
+        else
+            {
+            TR::Register *tempReg = cg->allocateRegister(TR_FPR);
+            TR_BackingStore *spillLocation = cg->allocateSpill(8, false, NULL);
+
+            TR::MemoryReference *spillRef = TR::MemoryReference::createWithSymRef(cg, node, spillLocation->getSymbolReference(), 8);
+            TR::MemoryReference *spillRefHi = TR::MemoryReference::createWithMemRef(cg, node, *spillRef, 0, 4);
+            TR::MemoryReference *spillRefLo = TR::MemoryReference::createWithMemRef(cg, node, *spillRef, 4, 4);
+
+            generateTrg1MemInstruction(cg, TR::InstOpCode::lfd, node, tempReg, memRef);
+            generateMemSrc1Instruction(cg, TR::InstOpCode::stfd, node, spillRef, tempReg);
+
+            if (cg->comp()->target().isSMP())
+                generateInstruction(cg, cg->comp()->target().cpu.isAtLeast(OMR_PROCESSOR_PPC_P7) ? TR::InstOpCode::lwsync : TR::InstOpCode::isync, node);
+
+            generateTrg1MemInstruction(cg, TR::InstOpCode::lwz, node, trgReg->getHighOrder(), spillRefHi);
+            generateTrg1MemInstruction(cg, TR::InstOpCode::lwz, node, trgReg->getLowOrder(), spillRefLo);
+
+            cg->freeSpill(spillLocation, 8, 0);
+            cg->stopUsingRegister(tempReg);
+            }
+        }
+    else
+#endif
+        {
+        TR::MemoryReference *memRefHi = TR::MemoryReference::createWithMemRef(cg, node, *memRef, 0, 4);
+        TR::MemoryReference *memRefLo = TR::MemoryReference::createWithMemRef(cg, node, *memRef, 4, 4);
+
+        generateTrg1MemInstruction(cg, TR::InstOpCode::lwz, node, trgReg->getHighOrder(), memRefHi);
+        generateTrg1MemInstruction(cg, TR::InstOpCode::lwz, node, trgReg->getLowOrder(), memRefLo);
+        }
+    }
+
+#ifdef J9_PROJECT_SPECIFIC
+enum class StoreSyncRequirements
+    {
+    None,
+    AtomicWrite,
+    LazySet,
+    Full
+    };
+
+StoreSyncRequirements getStoreSyncRequirements(TR::CodeGenerator *cg, TR::Node *node)
+    {
```

Comment:
Based on the existing comment ... at least, doxygen comment on this routine or clarify:  it only needs to be called for 32bit codegen, and you will not fall into the situation that isSyncVolatile is going to be a false-positive (i.e. side-effect of unresolved). Otherwise, change the comment appropriately.

Reply:
I've clarified the comment you were referring to. This method is meant to handle both 32-bit and 64-bit. The special `AtomicWrite` requirement when storing a volatile in non-SMP mode is meant to communicate to the caller that the write needs to be performed in one instruction. The comment was referring to the fact that this causes a different sequence to be used when storing longs on ppc32, but returning `AtomicWrite` is correct on 64-bit as well: outside of stores to longs on ppc32, `AtomicWrite` is simply equivalent to `None` as those stores are already done in one instruction.